sdk/packet/src/lib.rs

```diff
@@ -224,6 +292,61 @@ impl PartialEq for Packet {
     }
 }
 
+/// A custom implementation of io::Write to facilitate safe (non-UB)
+/// initialization of a MaybeUninit<Packet>
+struct PacketWriter {
+    // A pointer to the current write position
+    position: *mut u8,
+    // The number of remaining bytes that can be written to
+    spare_capacity: usize,
+}
+
+impl PacketWriter {
+    fn new_from_uninit_packet(packet: &mut mem::MaybeUninit<Packet>) -> Self {
+        // SAFETY: Access the field by pointer as creating a reference to
+        // and/or within the uninitialized Packet is undefined behavior
+        let position = unsafe { ptr::addr_of_mut!((*packet.as_mut_ptr()).buffer) as *mut u8 };
+        let spare_capacity = PACKET_DATA_SIZE;
+
+        Self {
+            position,
+            spare_capacity,
+        }
+    }
+
+    /// The offset of the write pointer within the buffer, which is also the
+    /// number of bytes that have been written
+    fn position(&self) -> usize {
+        PACKET_DATA_SIZE.saturating_sub(self.spare_capacity)
+    }
+}
+
+impl io::Write for PacketWriter {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if buf.len() > self.spare_capacity {
+            return Err(io::Error::from(io::ErrorKind::WriteZero));
+        }
+
+        // SAFETY: We previously verifed that buf.len() <= self.spare_capacity
+        // so this write will not push us past the end of the buffer. Likewise,
+        // we can update self.spare_capacity without fear of overflow
+        unsafe {
+            ptr::copy(buf.as_ptr(), self.position, buf.len());
```

Comment:
copy_nonoverlapping?

Reply:
Done in https://github.com/anza-xyz/agave/pull/3533/commits/085626429446011ceae15dc9f9ab392fbd23b08d